File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
import queue
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return []

        queue = deque([root])
        levels = []

        while len(queue):
            size = len(queue)
            level = []

            for _ in range(size):
                node = queue.popleft()

                level.append(node.val)

                if node.left != None:
                    queue.append(node.left)

                if node.right != None:
                    queue.append(node.right)

            levels.append(level)

        return levels

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
  vector<int> rightSideView(TreeNode *root) {
    vector<int> res;
    if (root == nullptr) {
      return res;
    }

    queue<TreeNode *> q;
    q.push(root);

    while (!q.empty()) {
      int size = q.size();

      vector<int> level;
      for (int i = 0; i < size; i++) {
        TreeNode *node = q.front();

        q.pop();
        if (node != nullptr) {
          level.push_back(node->val);

          if (node->right) {
            q.push(node->right);
          }

          if (node->left) {
            q.push(node->left);
          }
        }
      }
      res.push_back(level[0]);
    }

    return res;
  }
};
